### PR TITLE
Add custom path support

### PR DESCRIPTION
**Pull request by migmartri**
Add supports to deploy kubeapps under a custom prefix. 

To achieve it without requiring the user to do a custom build via webpack I needed to make the following modifications:

* In the package.json, set the `homepage` value relative so the bundled index.html contains a relative path to the assets.

```html
# Webpack generated index.html
# Before
<script type="text/javascript" src="/static/js/main.c4bfc712.js">
# after
<script type="text/javascript" src="./static/js/main.c4bfc712.js">
```

* Change React router history from push-state to hash-based. This simplified the setup since relative API calls defined in our code will ignore this hash state. This is not the case if we decided to modify the  push-state `basename` https://github.com/kubeapps/kubeapps/issues/402#issuecomment-448042009. Actually changing the basename would have been a runtime configuration that depends on the chart which seems brittle to me.  

* Updated Nginx rewrites and locations in both frontend and dashboard to handle the existence of an arbitrary namespace.

I could have done a rewrite in the Nginx target `nginx.ingress.kubernetes.io/rewrite-target` but I think that we are not requiring people to use only Nginx Ingress controller.

* Updated our dashboard code to use relative paths. 

peek 2018-12-18 17-26

### Call for testers

I have backed the dashboard in this docker image `migmartri/kubeapps-dashboard:devel16` and I'd appreciate if you could give it a try both deploying it with custom options but also tweaking the ingress path.

```
helm install --set dashboard.image.repository=migmartri/kubeapps-dashboard --set dashboard.image.tag=devel16 .
```

NOTE: 

In future PRs I will add the following annotation in the Ingress which will ensure (for users using Nginx ingress controller) that there is always a trailing slash

```
nginx.ingress.kubernetes.io/configuration-snippet:  rewrite ^(/[PREFIX])$ $1/ permanent;
```

Fixes https://github.com/kubeapps/kubeapps/issues/402

### CODE REVIEW

File: chart/kubeapps/templates/dashboard-config.yaml [hunk 1]
```diff
@@ -17,6 +17,15 @@ data:
       gzip_static  on;
 
       location / {
+        # Support for ingress prefixes maintaining compatibility with the default /
+        # 1 - Exactly two fragment URLs for files existing inside of the public/ dir
+        # i.e /[prefix]/config.json => /config.json
+        rewrite ^/[^/]+/([^/]+)$ /$1 break;
+
+        # 2 - Any static files bundled by webpack
+        # i.e /[prefix]/static/main.js => static.main.js
```
Comment (andresmgot) [inline chart/kubeapps/templates/dashboard-config.yaml]: either the comment is wrong or I am not understanding the regex, wouldn't this be:
```
         # i.e /[prefix]/static/main.js => /static/main.js
```
Also, isn't this regex already caught in the first?

File: dashboard/yarn.lock [hunk 2]
```diff
@@ -382,6 +382,14 @@ accepts@~1.3.4:
     mime-types "~2.1.16"
```
Comment (andresmgot) [inline dashboard/yarn.lock]: why did you need to update the dependencies? (I am asking because the package.json remains the same)

File: chart/kubeapps/templates/dashboard-config.yaml (hunk 1 above)
Comment (migmartri) [reply chart/kubeapps/templates/dashboard-config.yaml]: This comment is wrong, thanks for catching it up!

File: dashboard/yarn.lock (hunk 2 above)
Comment (migmartri) [reply dashboard/yarn.lock]: Good catch, this was because in previous versions of the patch I upgraded react-scripts-ts and for some reason yarn.lock did not downgrade. I'll put it back.

Thank you!

File: chart/kubeapps/templates/dashboard-config.yaml [hunk 3]
```diff
@@ -17,6 +17,15 @@ data:
       gzip_static  on;
 
       location / {
+        # Support for ingress prefixes maintaining compatibility with the default /
+        # 1 - Exactly two fragment URLs for files existing inside of the public/ dir
+        # i.e /[prefix]/config.json => /config.json
```
[the remark this answers is not in the document]
Comment (migmartri) [reply chart/kubeapps/templates/dashboard-config.yaml]: In kubectl proxy those two rewrites will be ignored as expected since they only match prefix defines in an ingress

In any case kubectl proxy does not work because of the header trim limitation.
 
selection_795

If you mean supporting multiple levels prefixes like /foo/kubeapps/ in the ingress, we do not support it and I think it's fine.

That scenario could be easily fixed if instead of providing all the files in `public` in the root i.e my-instance.com/favicon.png we add a prefix in a similar way we do with `static` `instance.com/public/favicon.png`

Having those two well defined directories would help with rewrites in the future.

**Comment by prydonius** [comment]
> In future PRs I will add the following annotation in the Ingress which will ensure (for users using Nginx ingress controller) that there is always a trailing slash

> nginx.ingress.kubernetes.io/configuration-snippet:  rewrite ^(/[PREFIX])$ $1/ permanent;

I just ran into this issue in my test, I was confused why it wasn't working for me, but I see you have addressed this in #897.
